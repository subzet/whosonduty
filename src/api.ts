import express from 'express';
import { createServer } from 'http';

import router from './controller';
import { environment } from './util';

const rest = () => {
  const app = express();

  app.set('trust proxy', 1);
  app.set('view cache', false);
  app.set('view engine', 'mustache');
  app.use(express.urlencoded({ extended: true }));

  app.use(express.json());
  app.get('/health-check', async (_, response) => {
    return response.json({ ok: true });
  });

  app.use('/', router);

  return app;
};

export async function server() {
  const app = rest();

  return createServer(app);
}

export async function start(): Promise<void> {
  const httpServer = await server();

  httpServer.listen(environment.PORT, '0.0.0.0', () => {
    console.info(`ðŸš€ REST http://localhost:${environment.PORT}`);
  });
}
