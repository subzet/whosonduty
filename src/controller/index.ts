import Router from 'express-promise-router';

import user from './user';

const router = Router();

router.use('/user', user);

export default router;
