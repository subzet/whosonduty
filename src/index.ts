export * from './api';
export * from './start';
