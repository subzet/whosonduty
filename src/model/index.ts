export * from './discord-channel';
export * from './incident-thread';
export * from './rotation';
export * from './user';
