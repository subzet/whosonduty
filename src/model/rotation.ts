import { getModelForClass, modelOptions, prop } from '@typegoose/typegoose';

import { MongoBase } from './base';

@modelOptions({
  schemaOptions: { _id: false },
})
export class RotationDuty {
  @prop()
  userId!: string;
  @prop()
  startDate!: Date;
  @prop()
  endDate!: Date;
  @prop()
  length!: string;
}

export class Rotation extends MongoBase {
  @prop()
  name!: string; // Backend Support;

  @prop()
  length!: string; // 7 days;

  @prop()
  responsible!: string[]; // [Manu, Ari, Marcos]

  @prop()
  duty!: RotationDuty[];
  // [Last RotationDuty, Current Rotation Duty , Next RotationDuty] o [Current Rotation Duty, Next Rotation Duty]
}

export const RotationModel = getModelForClass(Rotation);

export type RotationData = Omit<Rotation, keyof MongoBase>;
