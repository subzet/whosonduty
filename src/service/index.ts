export * from './discord-channel';
export * from './discord-user';
export * from './rotation';
