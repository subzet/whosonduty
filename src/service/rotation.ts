import dayjs, { Dayjs } from 'dayjs';
import ms from 'ms';

import { Rotation, RotationData, RotationDuty } from '../model';
import { mongoRotationRepository } from '../repository';
import { Errors } from '../util';

class RotationService {
  private isValidRotation = (data: RotationData) => {
    const lengthRegex = /\d* days/;

    if (!lengthRegex.test(data.length)) {
      throw Errors.invalidOperation(
        'Invalid rotation length. Try `<number> days`'
      );
    }

    if (data.responsible.length === 0) {
      throw Errors.invalidOperation(
        'Responsible list should have at least one responsible'
      );
    }
  };

  private isValidStartDate = (date: string, length: string) => {
    const isValidDate = dayjs(date).isValid();

    if (!isValidDate) {
      throw Errors.invalidOperation('Invalid date. Start date is incorrect');
    }

    const isValidStartDate = dayjs()
      .subtract(ms(length), 'millisecond')
      .isBefore(date);

    if (!isValidStartDate) {
      throw Errors.invalidOperation(
        'Invalid date. Start date should be after current date minus rotation length'
      );
    }
  };

  private getPeriodDates = (date: string, length: string): Dayjs[] => {
    return [dayjs(date), dayjs(date).add(ms(length), 'millisecond')];
  };

  private isNextPeriod(startDate: string): boolean {
    return dayjs().isBefore(startDate);
  }

  private generateRotationDuty = (
    data: RotationData,
    startDate: string,
    responsibleIndex = 0
  ) => {
    const result: RotationDuty[] = [];

    const [rotationDutyStartDate, rotationDutyEndDate] = this.getPeriodDates(
      startDate,
      data.length
    );

    let rotationDuty: RotationDuty = {
      startDate: rotationDutyStartDate.toDate(),
      endDate: rotationDutyEndDate.toDate(),
      userId: data.responsible[responsibleIndex],
    };

    result.push({ ...rotationDuty });

    while (!this.isNextPeriod(dayjs(rotationDuty.startDate).toISOString())) {
      const [rotationDutyStartDate, rotationDutyEndDate] = this.getPeriodDates(
        dayjs(rotationDuty.endDate).toISOString(),
        data.length
      );

      const currentResponsibleIndex = data.responsible.indexOf(
        rotationDuty.userId
      );

      const hasNextResponsible =
        currentResponsibleIndex + 1 < data.responsible.length;

      rotationDuty = {
        startDate: rotationDutyStartDate.toDate(),
        endDate: rotationDutyEndDate.toDate(),
        userId: hasNextResponsible
          ? data.responsible[currentResponsibleIndex + 1]
          : data.responsible[0],
      };

      result.push({ ...rotationDuty });
    }

    return result;
  };

  public create(
    data: RotationData,
    startDate = dayjs().startOf('day').toISOString()
  ): Promise<Rotation> {
    this.isValidRotation(data);

    this.isValidStartDate(startDate, data.length);

    const duty = this.generateRotationDuty(data, startDate);

    return mongoRotationRepository.create({ ...data, duty });
  }

  private shouldRotate = (rotation: Rotation) => {
    return dayjs().isAfter(rotation.duty?.[0].endDate);
  };

  public async rotate(rotationId: string) {
    const rotation = await mongoRotationRepository.findById(rotationId);

    if (!rotation.duty?.[0]) {
      throw Errors.invalidOperation('Rotation is invalid, no duties assigned');
    }

    if (!this.shouldRotate(rotation)) {
      throw Errors.invalidOperation(
        'Rotation can not be rotated because current rotation has not finished yet.'
      );
    }

    const currentRotationDuty = rotation.duty?.[0];

    const currentResponsibleIndex = rotation.responsible.indexOf(
      currentRotationDuty.userId
    );

    const hasNextResponsible =
      currentResponsibleIndex + 1 < rotation.responsible.length;

    const duty = hasNextResponsible
      ? this.generateRotationDuty(
          rotation,
          dayjs(currentRotationDuty.endDate).toISOString(),
          currentResponsibleIndex + 1
        )
      : this.generateRotationDuty(
          rotation,
          dayjs(currentRotationDuty.endDate).toISOString()
        );

    await mongoRotationRepository.update(rotationId, { ...rotation, duty });
  }

  public findById = async (rotationId: string) => {
    return mongoRotationRepository.findById(rotationId);
  };
}

export const rotationService = new RotationService();
