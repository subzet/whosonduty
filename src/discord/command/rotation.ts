import { Interaction, Message } from 'discord.js';

import { discordChannelService } from '../../service/discord-channel';
import { Errors } from '../../util';
import { DiscordCommand } from '../types';

const awaitInput = (interaction: Interaction) => {
  const filter = (message: Message) => {
    return message.author.id === interaction.user.id;
  };

  return interaction.channel?.awaitMessages({
    filter,
    max: 1,
    time: 10_000,
    errors: ['time'],
  });
};

const isValidParticipantsAnswer = (message: Message) => {
  if (message.mentions.users.size === 0) {
    throw Errors.invalidOperation(`No mentions in message`);
  }
};

const getParticipantsFromMessage = (message: Message) => {
  return message.mentions.users.map((user) => user.id);
};

export const rotation: DiscordCommand = {
  name: 'rotation',
  description: 'Creates a channel rotation',
  handler: async (interaction: Interaction) => {
    if (!interaction.isCommand()) {
      return;
    }

    try {
      await interaction.reply({ content: `Insert rotation participants` });

      const partipants = await awaitInput(interaction);

      const participantsAnswer = partipants?.first();

      if (!participantsAnswer) {
        throw Errors.invalidOperation();
      }

      isValidParticipantsAnswer(participantsAnswer);

      const participantsIds = getParticipantsFromMessage(participantsAnswer);

      await participantsAnswer?.reply({ content: `Insert rotation length` });

      const length = await awaitInput(interaction);

      const lengthAnswer = length?.first();

      const rotation = await discordChannelService.findOrCreateChannelRotation(
        interaction.channelId,
        'fake name',
        participantsIds
      );

      await interaction.channel?.send(
        `Finished creating rotation: ${JSON.stringify(rotation)}`
      );
    } catch {
      await interaction.channel?.send(
        `Did not receive any valid response, start over`
      );
    }

    const channelRotation =
      await discordChannelService.findOrCreateChannelRotation(
        interaction.channelId,
        'Rotation',
        ['1', '2', '3']
      );

    interaction.reply(
      `Channel rtation created ${JSON.stringify(channelRotation)}`
    );
  },
};
