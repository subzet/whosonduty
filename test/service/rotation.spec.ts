import dayjs from 'dayjs';
import ms from 'ms';

import { RotationData, RotationModel } from '../../src/model';
import { rotationService } from '../../src/service/rotation';
import { initDisposableDatabase } from '../helper';

let database: () => void;

async function createMockRotation(
  participants: number,
  firstDutyStartDate: string,
  currentIndex = 0,
  nextIndex = 1
) {
  // eslint-disable-next-line unicorn/no-new-array, unicorn/new-for-builtins
  const responsible = [...Array(participants)].map((_, index) =>
    String(index + 1)
  );

  const length = '5 days';

  return RotationModel.create({
    name: 'Fake',
    responsible,
    length,
    duty: [
      {
        userId: responsible[currentIndex],
        startDate: dayjs(firstDutyStartDate).toDate(),
        endDate: dayjs(firstDutyStartDate)
          .add(ms(length), 'millisecond')
          .toDate(),
      },
      {
        userId: responsible[nextIndex],
        startDate: dayjs(firstDutyStartDate)
          .add(ms(length), 'millisecond')
          .toDate(),
        endDate: dayjs(firstDutyStartDate)
          .add(ms(length) * 2, 'millisecond')
          .toDate(),
      },
    ],
  });
}

describe('Rotation test', () => {
  beforeAll(async () => {
    database = await initDisposableDatabase();
  });

  afterAll(async () => {
    database();
  });

  it('Should create rotation with participants', async () => {
    const data: RotationData = {
      name: 'Support',
      length: '5 days',
      responsible: ['1', '2', '3'],
    };

    const rotation = await rotationService.create(data);

    expect(rotation).toMatchSnapshot();
  });

  it('Should rotate participants', async () => {
    const mockRotation = await createMockRotation(
      3,
      dayjs().subtract(6, 'day').toISOString()
    );

    await rotationService.rotate(mockRotation._id.toString());

    const rotation = await rotationService.findById(
      mockRotation._id.toString()
    );

    expect(mockRotation).toMatchSnapshot();
    expect(rotation).toMatchSnapshot();
    expect(mockRotation.duty?.[1].userId).toBe(rotation.duty?.[0].userId);
  });

  it('Should rotate participants and begin again', async () => {
    const mockRotation = await createMockRotation(
      3,
      dayjs().subtract(6, 'day').toISOString(),
      2,
      0
    );

    await rotationService.rotate(mockRotation._id.toString());

    const rotation = await rotationService.findById(
      mockRotation._id.toString()
    );

    expect(mockRotation).toMatchSnapshot();
    expect(rotation).toMatchSnapshot();
    expect(mockRotation.duty?.[1].userId).toBe(rotation.duty?.[0].userId);
  });
});
