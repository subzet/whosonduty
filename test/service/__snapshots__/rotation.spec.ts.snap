// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rotation test Should create rotation with custom date 1`] = `
Object {
  "__v": 0,
  "_id": "6302ab599bad01ee7f552366",
  "createdAt": 2022-08-21T22:02:01.402Z,
  "duty": Array [
    Object {
      "endDate": 2022-08-25T03:00:00.000Z,
      "startDate": 2022-08-20T03:00:00.000Z,
      "userId": "1",
    },
    Object {
      "endDate": 2022-08-30T03:00:00.000Z,
      "startDate": 2022-08-25T03:00:00.000Z,
      "userId": "2",
    },
  ],
  "length": "5 days",
  "name": "Support",
  "responsible": Array [
    "1",
    "2",
    "3",
  ],
  "updatedAt": 2022-08-21T22:02:01.401Z,
}
`;

exports[`Rotation test Should create rotation with participants 1`] = `
Object {
  "__v": 0,
  "_id": "6302ab599bad01ee7f55235e",
  "createdAt": 2022-08-21T22:02:01.189Z,
  "duty": Array [
    Object {
      "endDate": 2022-08-26T03:00:00.000Z,
      "startDate": 2022-08-21T03:00:00.000Z,
      "userId": "1",
    },
    Object {
      "endDate": 2022-08-31T03:00:00.000Z,
      "startDate": 2022-08-26T03:00:00.000Z,
      "userId": "2",
    },
  ],
  "length": "5 days",
  "name": "Support",
  "responsible": Array [
    "1",
    "2",
    "3",
  ],
  "updatedAt": 2022-08-21T22:02:01.188Z,
}
`;

exports[`Rotation test Should rotate participants 1`] = `
Object {
  "__v": 0,
  "_id": "6302ab599bad01ee7f552360",
  "createdAt": 2022-08-21T22:02:01.312Z,
  "duty": Array [
    Object {
      "endDate": 2022-08-20T22:02:01.307Z,
      "startDate": 2022-08-15T22:02:01.307Z,
      "userId": "1",
    },
    Object {
      "endDate": 2022-08-25T22:02:01.307Z,
      "startDate": 2022-08-20T22:02:01.307Z,
      "userId": "2",
    },
  ],
  "length": "5 days",
  "name": "Fake",
  "responsible": Array [
    "1",
    "2",
    "3",
  ],
  "updatedAt": 2022-08-21T22:02:01.312Z,
}
`;

exports[`Rotation test Should rotate participants 2`] = `
Object {
  "__v": 0,
  "_id": "6302ab599bad01ee7f552360",
  "createdAt": 2022-08-21T22:02:01.312Z,
  "duty": Array [
    Object {
      "endDate": 2022-08-25T22:02:01.307Z,
      "startDate": 2022-08-20T22:02:01.307Z,
      "userId": "2",
    },
    Object {
      "endDate": 2022-08-30T22:02:01.307Z,
      "startDate": 2022-08-25T22:02:01.307Z,
      "userId": "3",
    },
  ],
  "length": "5 days",
  "name": "Fake",
  "responsible": Array [
    "1",
    "2",
    "3",
  ],
  "updatedAt": 2022-08-21T22:02:01.357Z,
}
`;

exports[`Rotation test Should rotate participants and begin again 1`] = `
Object {
  "__v": 0,
  "_id": "6302ab599bad01ee7f552368",
  "createdAt": 2022-08-21T22:02:01.416Z,
  "duty": Array [
    Object {
      "endDate": 2022-08-20T22:02:01.412Z,
      "startDate": 2022-08-15T22:02:01.412Z,
      "userId": "3",
    },
    Object {
      "endDate": 2022-08-25T22:02:01.412Z,
      "startDate": 2022-08-20T22:02:01.412Z,
      "userId": "1",
    },
  ],
  "length": "5 days",
  "name": "Fake",
  "responsible": Array [
    "1",
    "2",
    "3",
  ],
  "updatedAt": 2022-08-21T22:02:01.415Z,
}
`;

exports[`Rotation test Should rotate participants and begin again 2`] = `
Object {
  "__v": 0,
  "_id": "6302ab599bad01ee7f552368",
  "createdAt": 2022-08-21T22:02:01.416Z,
  "duty": Array [
    Object {
      "endDate": 2022-08-25T22:02:01.412Z,
      "startDate": 2022-08-20T22:02:01.412Z,
      "userId": "1",
    },
    Object {
      "endDate": 2022-08-30T22:02:01.412Z,
      "startDate": 2022-08-25T22:02:01.412Z,
      "userId": "2",
    },
  ],
  "length": "5 days",
  "name": "Fake",
  "responsible": Array [
    "1",
    "2",
    "3",
  ],
  "updatedAt": 2022-08-21T22:02:01.435Z,
}
`;
